Add support for HTMLLabelElement.control

FrecksterGIT opened the pull request:
Adds support for the HTMLLabelElement.control and *.labels properties. (jsdom/jsdom#2175)

Adds labels property support for:

- button
- input (if the type attribute is not in the Hidden state)
- meter
- output
- progress
- select
- textarea

Activated tests:
- button-labels.html
- input-labels.html
- labelable-elements.html

Added tests:
- label-attributes-dont-upstream.html (copy from wpt without 2 failing tests.)

TimothyGu commented on lib/jsdom/living/nodes/HTMLLabelElement-impl.js:
```diff
@@ -48,6 +48,26 @@ function sendClickToAssociatedNode(node) {
+      if (node && isLabelable(node)) {
+        return node;
+      }
+      return undefined;
```
This should probably be null.

FrecksterGIT replied:
Fixed. Also changed the activation behaviour to make use of this. it didn't handle the _wrong-for-attribute-case_.

domenic commented:
Nice work! A lot of stuff about `control` seems to be covered by https://github.com/w3c/web-platform-tests/blob/master/html/semantics/forms/the-label-element/label-attributes.html. Can you enable that test and see how we do?

FrecksterGIT commented:
After working through the tests, I also changed the form to make use of the control.

Ignoring all tests using HTMLInputElement.labels the only one that is still failing:
```
Failed in "A label element not in a document can not label any element in the document.":
assert_not_equals: A label element not in a document should not label an element in a document. got disallowed value Element node <input id="test1"></input>
```
The problem here is, that the label has an ownerDocument, but isn't in the dom.

```javascript
    var label = document.createElement("label");
    label.htmlFor = "test1";
    assert_not_equals(label.control, document.getElementById("test1"),
                      "A label element not in a document should not label an element in a document.");
    document.body.appendChild(label);
```

Seems more like a generic problem with creating elements and adding/removing them from the dom.

TimothyGu commented on lib/jsdom/living/nodes/HTMLLabelElement-impl.js:
```diff
@@ -48,27 +47,35 @@ function sendClickToAssociatedNode(node) {
       const node = this.ownerDocument.getElementById(this.getAttribute("for"));
-      if (node && isLabelable(node) && !isDisabled(node)) {
-        sendClickToAssociatedNode(node);
+      if (node && isLabelable(node)) {
```
The spec [says](https://html.spec.whatwg.org/multipage/forms.html#attr-label-for):

> If the [[`for`](https://html.spec.whatwg.org/multipage/forms.html#attr-label-for)] attribute is specified and there is an element in the [tree](https://dom.spec.whatwg.org/#concept-tree) whose [ID](https://dom.spec.whatwg.org/#concept-id) is equal to the value of the [`for`](https://html.spec.whatwg.org/multipage/forms.html#attr-label-for) attribute, and the first such element in [tree order](https://dom.spec.whatwg.org/#concept-tree-order) is a [labelable element](https://html.spec.whatwg.org/multipage/forms.html#category-label), then that element is the [`label`](https://html.spec.whatwg.org/multipage/forms.html#the-label-element) element's [labeled control](https://html.spec.whatwg.org/multipage/forms.html#labeled-control).

You need to check if the current element is in the same tree as node. A simple algorithm would be

```js
function hasCommonAncestor(a, b) {
  const ancestors = new Set(domSymbolTree.ancestorsIterator(a));
  for (const ancestor of domSymbolTree.ancestorsIterator(b)) {
    if (ancestors.has(ancestor)) {
      return true;
    }
  }
  return false;
}
```

This should fix the remaining test failure.

FrecksterGIT commented:
Added support for HTMLInputElement.labels.
All control-attribute related tests are passing now.

Only got 3 failing tests left in label-attributes.html.

TimothyGu commented:
@FrecksterGIT can you enable the passing `control` attribute tests in this pull request as well?

TimothyGu commented on lib/jsdom/living/nodes/HTMLLabelElement-impl.js:
```diff
@@ -29,6 +28,11 @@ function isLabelable(node) {
 
+function hasCommonAncestor(a, b) {
+  const ancestors = new Set(domSymbolTree.ancestorsIterator(a));
+  return domSymbolTree.ancestorsToArray(b).some(ancestor => ancestors.has(ancestor));
```
Can you use `ancestorsIterator` as I originally proposed? That would obviate the need to walk through the entire ancestor chain for `b` if not needed.

TimothyGu commented on lib/jsdom/living/nodes/HTMLInputElement-impl.js:
```diff
@@ -316,6 +320,16 @@ class HTMLInputElementImpl extends HTMLElementImpl {
+      query: () => domSymbolTree.treeToArray(root, {
+        filter: node => node.tagName === "LABEL" && node.control === this
+      })
+    });
```
Please always implement things according to the [spec](https://html.spec.whatwg.org/multipage/forms.html#dom-lfe-labels), which specifically says that the same object must be returned on separate invocations of `labels` getter, and that the `labels` getter should return null in case the type of this `input` element is Hidden.

FrecksterGIT commented:
```
html/semantics/forms/the-input-element
      √ input-labels.html
    html/semantics/forms/the-label-element
      1) label-attributes.html
```
Still two failing tests in label-attributes.html:

1. document.getElementById(...).attachShadow is not a function

> Web Components / Shadow Dom isn't supported by jsdom.

2. Failed in "A labelable element is moved to iframe.":
assert_equals: The number of labels should be 2 after the labelable element is moved to iframe. expected 2 but got 0

> not sure how to tackle this one.

I haven't figured out a way to disable single wpt-tests in one test-file. Guess it's not even possible?
And is there some documentation for debugging wpt? That took longer than it should have.

FrecksterGIT commented:
Refactored labels logic into helpers/form-controls.js and added labels support to all labelable elements i could find:

https://html.spec.whatwg.org/multipage/forms.html#category-label

- button
- input (if the type attribute is not in the Hidden state)
- meter
- output
- progress
- select
- textarea

```
  web-platform-tests html/semantics/forms/the-label-element labelable-elements.html:
     Error: Failed in "Check if the keygen element is not a labelable element": 
assert_not_equals: got disallowed value Element node <keygen id="testkeygen"></keygen>
```
Keygen is deprecated and not part of the living spec.

TimothyGu commented on lib/jsdom/living/nodes/HTMLLabelElement-impl.js:
```diff
@@ -2,31 +2,16 @@
-  // labelable logic defined at: https://html.spec.whatwg.org/multipage/forms.html#category-label
-  if (node.nodeType !== 1) {
-    return false;
+function getRootNode(node) {
```
You can use the same function in traversal.js.

FrecksterGIT replied:
I knew i missed something there.

TimothyGu commented on lib/jsdom/living/helpers/form-controls.js:
```diff
@@ -60,6 +64,50 @@ exports.normalizeToCRLF = string => {
+  if (!exports.isLabelable(labelable)) {
+    return null;
+  }
+  if (!labelable._labels) {
```
Can you declare `this._labels = undefined` in the constructors of all classes that use this function?

FrecksterGIT replied:
Initialized `this._labels = null` everywhere.

TimothyGu commented:
It also seems like there are some outstanding test failures: https://travis-ci.org/jsdom/jsdom/jobs/381198906

Some of the failures seem to have to do with `attachShadow` which we don't implement. In that case, you can copy the file from upstream into our to-upstream folder, commenting out the features we don't yet implement, and change the filename so that it has `dont-upstream` in the name. Like [this](https://github.com/jsdom/jsdom/blob/master/test/web-platform-tests/to-upstream/hr-time/timeOrigin-dont-upstream.html).
